save-core/src/commonMain/kotlin/org/cqfn/save/core/files/MergeConfigs.kt

```diff
@@ -0,0 +1,112 @@
+package org.cqfn.save.core.files
+
+import org.cqfn.save.core.config.TestConfig
+import org.cqfn.save.core.logging.logDebug
+import org.cqfn.save.core.plugin.GeneralConfig
+import org.cqfn.save.plugin.warn.WarnPluginConfig
+import org.cqfn.save.plugins.fix.FixPluginConfig
+
+/**
+ * A class that is capable for merging inherited configurations
+ */
+class MergeConfigs {
+    /**
+     * Merge parent configurations with current and prolong it for all child configs
+     *
+     * @param testConfig - testing SAVE config (save.toml) which should be merged
+     */
+    fun merge(testConfig: TestConfig) {
+        logDebug("Start merging configs for ${testConfig.location}")
+        val parentConfigs = collectParentConfigs(testConfig)
+        mergeConfigList(parentConfigs)
+        mergeChildConfigs(testConfig)
+    }
+
+    // Create the list of parent configs
+    private fun collectParentConfigs(testConfig: TestConfig): MutableList<TestConfig> {
+        val configList = mutableListOf(testConfig)
+        var parentConfig = testConfig.parentConfig
+        while (parentConfig != null) {
+            configList.add(parentConfig)
+            parentConfig = parentConfig.parentConfig
+        }
+        configList.reverse()
+        return configList
+    }
+
+    // Merge list of configs pairwise
+    private fun mergeConfigList(configList: MutableList<TestConfig>) {
+        if (configList.size == 1) {
+            return
+        }
+        val pairs = configList.zipWithNext()
+
+        pairs.forEach { (parent, child) ->
+            mergeChildConfigWithParent(parent, child)
+        }
+    }
+
+    // Merge child configs recursively
+    private fun mergeChildConfigs(testConfig: TestConfig) {
+        for (child in testConfig.childConfigs) {
+            mergeChildConfigWithParent(testConfig, child)
+            mergeChildConfigs(child)
+        }
+    }
+
+    private fun mergeChildConfigWithParent(parent: TestConfig, child: TestConfig) {
+        logDebug("Merging ${parent.location} with ${child.location}")
+        val parentConfig = parent.pluginConfigs
+        val childConfig = child.pluginConfigs
+        // Create the list of corresponding configs, if one of them will be null -> list will contain only one element,
+        // which we apply as final config, otherwise we will merge configs
+        val generalConfigs = listOfNotNull(
+            parentConfig.filterIsInstance<GeneralConfig>().firstOrNull(),
+            childConfig.filterIsInstance<GeneralConfig>().firstOrNull()
+        )
+        val newGeneralConfig: GeneralConfig? = if (generalConfigs.size != 2) generalConfigs.firstOrNull() else mergeGeneralConfigs(generalConfigs.first(), generalConfigs.last())
```

Comment:
```suggestion
        val newGeneralConfig: GeneralConfig? = generalConfigs.singleOrNull() ?: mergeGeneralConfigs(generalConfigs.first(), generalConfigs.last())
```

Reply:
Basically, I need to consider three cases:
1) If list is empty -> return null
2) If `list.size == 1` -> return single element
3) If `list.size == 2` -> return merge elements
-----------------------
But your suggestion will give the following:
1) If list is empty -> singleOrNull returns null -> mergeConfgis -- ERROR
2) If `list.size == 1` -> return single element -- OK
3) If `list.size == 2` -> return merge elements -> singleOrNull returns null -> mergeConfgis -- OK

Reply:
I think, the reason was in unclear comment before this code, I provided more detailed explanation